src/v/pandaproxy/schema_registry/store.h

```diff
@@ -19,6 +19,23 @@
 
 namespace pandaproxy::schema_registry {
 
+namespace detail {
+
+template<typename T>
+typename T::iterator
+make_non_const_iterator(T& container, typename T::const_iterator it) {
+    return container.erase(it, it);
+}
+
+template<typename T>
+result<typename T::iterator>
+make_non_const_iterator(T& container, result<typename T::const_iterator> it) {
+    auto res = BOOST_OUTCOME_TRYX(it);
```

Comment:
can we add similar functionality as *our* macro. you can redefine it if you want, but I think it would be better for us to do it ourselves. 

Also, wonder if we can do the same w/ templates and constexpr things rather than macros given that we are in c++20. typically macros in boost were bridging features for later c++

Reply:
> can we add similar functionality as *our* macro. you can redefine it if you want, but I think it would be better for us to do it ourselves. 
> 
> Also, wonder if we can do the same w/ templates and constexpr things rather than macros given that we are in c++20. typically macros in boost were bridging features for later c++

C++ doesn't allow conditional return, other than via exceptions, which `result<T>` aims to avoid. It might be simpler to just use `ss::future<T>`. If I figure it out, I'll be sure to write a blog post on monads.

Reply:
@senior7515 Any suggestions on the name? Or I could just switch `store.h` to `ss::future` and avoid the translation altogether.

Reply:
> interestingly this particular macro is hardly a macro, its an expression statement https://www.boost.org/doc/libs/1_70_0/libs/outcome/doc/html/tutorial/essential/result/try.html which is a compiler extension in gcc/clang. so I think that it doesn't quite work with generic language features and still needs to be a macro.

```cpp
 auto res = BOOST_OUTCOME_TRYX(it);
```
Sort of expands like this:
```cpp
auto res_unique_name = it;
if(res_unique_name.has_error()) {
   return std::move(res_unique_name).error();
}
auto res = std::move(res_unique_name).value();
```

Reply:
i think `utils::exception_as_result()` or even a static func 

```
template<typename T>
consteval
result<T>
result::wrapx() noexcept;
```

Reply:
> i think `utils::exception_as_result()` or even a static func
> 
> ```
> template<typename T>
> consteval
> result<T>
> result::wrapx() noexcept;
> ```

I think that's the opposite conversion than whats needed?

I might be able to get away with just calling `.value()`, the exception is handled at the top of the handler anyway. At that point I might just replace `result` with `ss::future` as the "expected" type, since it ends up as an exception in a lot of places, anyway, and logging the stack will have more info.